src/modules/bulletproofs/tests_impl.h

```diff
@@ -0,0 +1,225 @@
+    CHECK(secp256k1_bulletproofs_pp_log2(2) == 1);
+    CHECK(secp256k1_bulletproofs_pp_log2(255) == 7);
+    CHECK(secp256k1_bulletproofs_pp_log2(256) == 8);
+    CHECK(secp256k1_bulletproofs_pp_log2(257) == 8);
+}
+
+void test_norm_util_helpers(void) {
+    secp256k1_scalar a_vec[4], b_vec[4], r_pows[4], res, res2, q, r;
+    int i;
+    /* a = {1, 2, 3, 4} b = {5, 6, 7, 8}, q = 4, r = 2 */
+    for (i = 0; i < 4; i++) {
+        secp256k1_scalar_set_int(&a_vec[i], i + 1);
+        secp256k1_scalar_set_int(&b_vec[i], i + 5);
+    }
+    secp256k1_scalar_set_int(&q, 4);
+    secp256k1_scalar_set_int(&r, 2);
+    secp256k1_scalar_inner_product(&res, a_vec, 0, b_vec, 0, 1, 4);
+    secp256k1_scalar_set_int(&res2, 70);
+    CHECK(secp256k1_scalar_eq(&res2, &res) == 1);
+
+    secp256k1_scalar_inner_product(&res, a_vec, 0, b_vec, 1, 2, 2);
+    secp256k1_scalar_set_int(&res2, 30);
+    CHECK(secp256k1_scalar_eq(&res2, &res) == 1);
+
+    secp256k1_scalar_inner_product(&res, a_vec, 1, b_vec, 0, 2, 2);
+    secp256k1_scalar_set_int(&res2, 38);
+    CHECK(secp256k1_scalar_eq(&res2, &res) == 1);
+
+    secp256k1_scalar_inner_product(&res, a_vec, 1, b_vec, 1, 2, 2);
+    secp256k1_scalar_set_int(&res2, 44);
+    CHECK(secp256k1_scalar_eq(&res2, &res) == 1);
+
+    secp256k1_weighted_scalar_inner_product(&res, a_vec, 0, a_vec, 0, 1, 4, &q);
+    secp256k1_scalar_set_int(&res2, 4740); /*i*i*4^(i+1) */
+    CHECK(secp256k1_scalar_eq(&res2, &res) == 1);
+
+    secp256k1_bulletproofs_powers_of_r(r_pows, &r, 4);
+    secp256k1_scalar_set_int(&res, 2); CHECK(secp256k1_scalar_eq(&res, &r_pows[0]));
+    secp256k1_scalar_set_int(&res, 4); CHECK(secp256k1_scalar_eq(&res, &r_pows[1]));
+    secp256k1_scalar_set_int(&res, 16); CHECK(secp256k1_scalar_eq(&res, &r_pows[2]));
+    secp256k1_scalar_set_int(&res, 256); CHECK(secp256k1_scalar_eq(&res, &r_pows[3]));
+}
+
+void norm_arg_test(unsigned int n, unsigned int m) {
+    secp256k1_scalar n_vec[64], l_vec[64], c_vec[64];
+    secp256k1_scalar r, q;
+    secp256k1_ge commit;
+    size_t i, plen;
+    int res;
+    secp256k1_bulletproofs_generators *gs = secp256k1_bulletproofs_generators_create(ctx, n + m);
+    secp256k1_scratch *scratch = secp256k1_scratch_space_create(ctx, 1000*1000); /* shouldn't need much */
+    unsigned char proof[1000], transcript32[32];
+    for (i = 0; i < 32; i++) {
+        transcript32[i] = i;
+    }
+    random_scalar_order(&r);
+    secp256k1_scalar_sqr(&q, &r);
+
+    for (i = 0; i < n; i++) {
+        random_scalar_order(&n_vec[i]);
+    }
+
+    for (i = 0; i < m; i++) {
+        random_scalar_order(&l_vec[i]);
+        random_scalar_order(&c_vec[i]);
+    }
+
+    res = secp256k1_bulletproofs_commit(ctx, scratch, &commit, gs, n_vec, n, l_vec, m, c_vec, m, &q);
+    CHECK(res == 1);
+    res = secp256k1_bulletproofs_pp_rangeproof_norm_product_prove(ctx, scratch, proof, &plen, transcript32, &r, gs, n_vec, n, l_vec, m, c_vec, m, &commit);
```

Comment:
Ok I found it: `plen` is not initialized but read at the start of the `norm_product_prove` function. When I initialize plen as `sizeof(proof)` the tests succeed.